generator/src/AbstractGenerator.php

```diff
@@ -0,0 +1,127 @@
+<?php
+declare(strict_types=1);
+
+namespace MongoDB\CodeGenerator;
+
+use InvalidArgumentException;
+use MongoDB\Builder\Expression;
+use MongoDB\CodeGenerator\Definition\ArgumentDefinition;
+use MongoDB\CodeGenerator\Definition\GeneratorDefinition;
+use Nette\PhpGenerator\ClassType;
+use Nette\PhpGenerator\PhpFile;
+use Nette\PhpGenerator\Printer;
+use Nette\PhpGenerator\PsrPrinter;
+
+use function dirname;
+use function file_put_contents;
+use function implode;
+use function in_array;
+use function interface_exists;
+use function is_dir;
+use function mkdir;
+use function sort;
+use function ucfirst;
+
+/** @internal */
+abstract class AbstractGenerator
+{
+    /** @var array<string, list<string|class-string>> */
+    protected array $typeAliases = [
+        'resolvesToExpression' => [Expression\ResolvesToExpression::class, 'array', 'object', 'string', 'int', 'float', 'bool', 'null'],
+        'resolvesToArrayExpression' => [Expression\ResolvesToArrayExpression::class, 'array', 'object', 'string'],
+        'resolvesToBoolExpression' => [Expression\ResolvesToBoolExpression::class, 'array', 'object', 'string', 'bool'],
+        'resolvesToMatchExpression' => ['array', 'object', Expression\ResolvesToMatchExpression::class],
+        'resolvesToNumberExpression' => [Expression\ResolvesToBoolExpression::class, 'array', 'object', 'string', 'int', 'float'],
```

Comment:
Just a quick heads up that the type substitution rules have changed since I built the original prototype. Mainly, you can no longer use arrays, objects, or strings to pass in your own expressions or field paths: https://docs.google.com/document/d/1vFWXB_6nnYH6p136alKDnIHekW-JeNA7nH3vWSriDtY/edit#bookmark=id.3rypi8grpbbo

Reply:
I updated the expression type system, adding a config file to generate expression classes and interfaces.